src/API.jl

```diff
@@ -894,71 +894,89 @@ end
 
 precompile() = precompile(Context())
 function precompile(ctx::Context)
+    
+    function is_stale(paths, sourcepath)
+        toml_c = Base.TOMLCache()
+        for path_to_try in paths::Vector{String}
+            staledeps = Base.stale_cachefile(sourcepath, path_to_try, toml_c)
+            staledeps === true ? continue : return false
+        end
+        return true
+    end
```

Comment:
I suggest to move `is_stale` out of `precompile`. As of https://github.com/JuliaLang/Pkg.jl/pull/2021/commits/88b5a71873e641b3e23074de25d811265f98c805 we have

https://github.com/JuliaLang/Pkg.jl/blob/88b5a71873e641b3e23074de25d811265f98c805/src/API.jl#L941-L944

I think it makes `is_stale` a closure that has `toml_c` as a boxed value as in this toy example:

```julia
julia> function f()
           g() = toml_c = 1
           toml_c = 2
           g
       end;

julia> f().toml_c
Core.Box(2)
```

Since `is_stale` is used in `@async`, it would mean that all tasks share the same `Base.TOMLCache()`.

(We *could* avoid this by renaming `toml_c` to something else. But I think closures with `@async`/`@spawn` are very fragile and hard to maintain in general. (Although obviously closures are super useful.) So, I think defining it as a function is the best approach here.)

If it's not ideal to put `is_stale` in `Pkg.API` namespace, maybe use `_is_stale` or define it in another submodule?

Reply:
Indeed! Done. Thanks